more warning squash

pavanbalaji opened the pull request:
## Pull Request Description

This time on a linux build for ch4:ofi.

## Expected Performance Changes

None.

## Known Issues

Only looks at `ch4:ofi`.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Passes tests (included warning check)
* [x] Confirm whitespace/style checkers are happy (or has a good reason for being bad)
* [x] Commits are self-contained and do not do two things at once
* [x] Remove xfail from the test suite when fixing a test
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Add comments such that someone without knowledge of the code could understand
* [x] Add Devel Docs in the `doc/` directory for any new code design

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

hzhou commented:
@pavanbalaji What configure option did you use to see these warnings?

pavanbalaji commented:
> @pavanbalaji What configure option did you use to see these warnings?

`--enable-strict --enable-g=all`

hzhou commented:
> > @pavanbalaji What configure option did you use to see these warnings?
> 
> `--enable-strict --enable-g=all`

And what compiler and version? I am wondering why the warnings checker didn't catch these before...

pavanbalaji commented:
> And what compiler and version? I am wondering why the warnings checker didn't catch these before...

`gcc-4.6.3` (on the jenkins nodes)

hzhou commented:
I see. I was testing on the new Jenkins server, and that is using `gcc 4.8.5`.

pavanbalaji commented:
@hzhou Since you have been working on the warning squashes too, it might be better if you review this PR, rather than Ken.  (of course, he can review too, if he has time).

hzhou commented on src/mpid/ch4/netmod/ofi/ofi_rma.h:
```diff
@@ -876,7 +876,7 @@ static inline int MPIDI_OFI_do_accumulate(const void *origin_addr,
     int rc, mpi_errno = MPI_SUCCESS;
-    uint64_t flags;
+    uint64_t flags = 0;
     MPIDI_OFI_win_request_t *req = NULL;
```
I have been hesitating on this type of warning-fix. The compilers warns us when it thinks there is a code path lead to variable being used without set (I assume that's the warning). By blindly setting initial value, it is equivalent to disable that particular warning. I.E. in the case that we do have code logic error, the compiler will not warn us anymore. We are just replacing the potential use-uninitialized error with potential use-wrong-value error.

The proper fix, IMHO, is to address why the compiler failed to figure out the code path. Sometime this is difficult due to our legacy frame code or compiler being too old or too silly. When that is too difficult, I would suggest at least move the (often) temporary variable close to where it is being used and maybe also add a comment explaining the code path. That way, it is a less blind fix.

pavanbalaji replied:
That's a fair point and I agree.  Let me spend some more time on it.

pavanbalaji replied:
I'm not able to reproduce this warning at all.  Very strange.  I'll drop this part of the patch.

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

hzhou commented:
test:mpich/warnings/more

hzhou commented:
test:mpich/warnings/icc

hzhou commented:
test:mpich/icc/warnings

hzhou commented:
test:icc/warnings

hzhou commented:
I added `gcc-4.6` (using `gcc-4.6.3`) to the `more-warnings` test. There are still some warnings left for `ch4-ofi`, the other devices are good.

hzhou commented:
test:icc/warnings

pavanbalaji commented:
@hzhou If you are using an external libfabric, you'll need to patch it (see #3802) to remove the warnings.  The patches have already been submitted upstream:

https://github.com/ofiwg/libfabric/pull/5041
https://github.com/ofiwg/libfabric/pull/5042

hzhou commented:
test:icc/warnings

hzhou commented:
I updated the libfabric by pulling from `pmodels/libfabric` branch `v1.7.1-mpich`. There is still one warning remain:
```
/var/lib/jenkins-slave/workspace/hzhou-warnings-more/compiler/gcc-4.6/config/ch4-ofi/label/centos64/modules/libfabric/include/rdma/fi_atomic.h:175:2: warning: ‘flags’ may be used uninitialized in this function [-Wuninitialized]

```

pavanbalaji commented:
> I updated the libfabric by pulling from `pmodels/libfabric` branch `v1.7.1-mpich`. There is still one warning remain:
> 
> ```
> /var/lib/jenkins-slave/workspace/hzhou-warnings-more/compiler/gcc-4.6/config/ch4-ofi/label/centos64/modules/libfabric/include/rdma/fi_atomic.h:175:2: warning: ‘flags’ may be used uninitialized in this function [-Wuninitialized]
> ```

That was the one warning that I removed the "fix" for because I couldn't reproduce it anymore.  I'll try again.

hzhou commented:
Oh I wasn't realizing that wasn't our code :)  in that case I don't think I really care that much. It's up to you and the libfabric folks.

pavanbalaji commented:
It is our code.  I need to reproduce it again.  I wasn't able to do that last night.

hzhou commented:
I see. It seems the reason it trips `gcc 4.6` is `MPIDI_OFI_allocate_win_request_put_get` may `goto fn_fail` before `MPIDI_OFI_INIT_SIGNAL_REQUEST` call. May I suggest to pacify the warning by add 
```
*flags = 0; /* to pacify gcc-4.6 -Wuninitialized */
```
under `fn_fail:` label?

pavanbalaji commented:
That defeats your own advice and replaces "the potential use-uninitialized error with potential use-wrong-value error".  :-)  Let me look into the warning.  It might be a real problem.

pavanbalaji commented:
> > That defeats your own advice and replaces "the potential use-uninitialized error with potential use-wrong-value error". :-) Let me look into the warning. It might be a real problem.
> 
> We know for sure that at `fn_fail`the value of `flags` does not matter since the error will be poped all the way up, so there is no such potential. Any code outside the `fn_fail` code path is not compromised.

Depends on what we do when the function returns, so the `fn_fail` path could be compromised.  Let me look into the warning.  There's no point in arguing before I do that.

pavanbalaji commented:
This is definitely a compiler bug.  After the preprocessor parse, the code eventually gets transformed to something like this:

```
int foo(int *flag) {
    if (something_bad) {
        mpi_errno = MPIR_Err_create_code(...);
        goto fn_fail;
    }
    *flag = 100;
}

mpi_errno = foo(&flag);
if (mpi_errno) {
    goto fn_fail;
}

fi_atomicmsg(..., flag);
```

In this case, the compiler does not know that `MPIR_Err_create_code` will always return a non-zero error code, so it thinks that we might use flag in an undefined manner.

hzhou commented:
> This is definitely a compiler bug...

You cannot blame a compiler not intelligent enough as a bug :)

The newer versions of `gcc` is smarter.

pavanbalaji commented:
> > This is definitely a compiler bug...
> 
> You cannot blame a compiler not intelligent enough as a bug :)
> 
> The newer versions of `gcc` is smarter.

I guess it's just a warning, so the compiler doesn't have to be sure before it can throw it.  How new a version?

hzhou commented:
Even `gcc 4.8.5` does not throw that warning.

pavanbalaji commented:
test:jenkins/ch3/most
test:jenkins/ch4/most
test:mpich/warnings/more

hzhou commented:
The warnings are clean. Ignore the `pgi` tests, they are still being worked out.

pavanbalaji commented:
@hzhou This is ready for your review.

pavanbalaji commented on src/include/mpir_err.h:
```diff
@@ -810,28 +810,38 @@ void MPIR_Handle_fatal_error(struct MPIR_Comm *comm_ptr, const char fcname[], in
-                                    __LINE__, class_, msg_, 0); stmt_ ; \
+                                    __LINE__, class_, msg_, 0);         \
+        assert(err_);                                                   \
+        stmt_ ;                                                         \
```
I had added it in a previous version, but I removed it because the assert is generally useful even if the warning wasn't there.  It's an extra level of error checking to make sure someone doesn't accidentally use the error creation function incorrectly (e.g., by setting both the error code and error class as `MPI_SUCCESS`).

hzhou replied:
OK.
